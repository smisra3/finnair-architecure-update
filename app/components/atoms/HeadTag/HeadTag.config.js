export default {
  META_KEYS: [
    {
      key: 'author',
      id: 1,
    },
    {
      key: 'description',
      id: 2,
    },
  ],
  LINK: ['canonical'],
};
