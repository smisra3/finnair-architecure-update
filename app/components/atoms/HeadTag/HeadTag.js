// @flow

import Head from 'next/head';
import { PureComponent } from 'react';

import config from './HeadTag.config';

type Props = {
  title?: string,
  schema?: any,
};

type MetaProps = {
  prefix?: string,
  meta: {
    id: number,
    key: string,
    content: string,
  },
  content: string,
};

export const MetaTag = (props: MetaProps) => {
  const attrs = {
    content: props.content,
  };
  // eslint-disable-next-line
  attrs[props.prefix && props.prefix === 'og' ? 'property' : 'name'] = props.prefix
    ? `${props.prefix}:${props.meta.key}`
    : props.meta.key;

  return <meta {...attrs} />;
};

class HeadTag extends PureComponent<Props> {
  static defaultProps = {
    title: '',
    schema: undefined,
  };

  render() {
    const { title, schema } = this.props;
    const { META_KEYS, LINK } = config;
    return (
      <Head>
        {/* eslint-disable react/no-danger */}
        <title>{title}</title>
        {META_KEYS.map((meta) => {
          const content = this.props[meta.serverKey || meta.key];
          if (!content) {
            return null;
          }
          return meta.prefix ? (
            meta.prefix.map(prefix => (
              <MetaTag
                key={`meta-${prefix}-${meta.id}`}
                prefix={prefix}
                meta={meta}
                content={content}
              />
            ))
          ) : (
            <MetaTag key={`meta-${meta.id}`} meta={meta} content={content} />
          );
        })}
        <meta httpEquiv="Accept-CH" content="DPR, Width" />
        {LINK.map(key => (this.props[key]
          ? <link rel={key} href={this.props[key]} key={key} />
          : null))}

        {/* eslint-disable react/no-danger */}
        {schema ? (
          <script
            dangerouslySetInnerHTML={{
              __html: JSON.stringify(schema),
            }}
            type="application/ld+json"
          />
        ) : null}
        {/* eslint-enable react/no-danger */}
      </Head>
    );
  }
}

MetaTag.defaultProps = {
  prefix: undefined,
};

export default HeadTag;
