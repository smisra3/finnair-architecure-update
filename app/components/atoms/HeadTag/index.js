// @flow
export { default } from './HeadTag';
