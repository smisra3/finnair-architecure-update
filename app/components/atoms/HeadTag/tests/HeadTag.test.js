import { shallow } from 'enzyme';
import HeadTag from '../';
import { MetaTag } from '../HeadTag';

const sampleSiteData = {
  title: 'title',
  author: 'author',
  description: 'description',
  image: 'image',
  url: 'url',
  type: 'type',
  siteName: 'siteName',
  card: 'sample_description',
};

describe('HeadTag Component', () => {
  let head;
  const metaData = {
    content: 'content',
    prefix: 'prefix',
    meta: {
      key: 'meta key',
    },
  };

  test('should render correctly and match snapshot', () => {
    head = shallow(<HeadTag {...sampleSiteData} />);
    expect(head).toMatchSnapshot();
  });

  test('should render correctly when no props passed', () => {
    head = shallow(<HeadTag />);
    expect(head).toMatchSnapshot();
  });

  test('should render tags for only the props that were passed', () => {
    head = shallow(<HeadTag title="mytitle" />);
    expect(head.contains(<title>mytitle</title>)).toEqual(true);
    expect(head.find('meta')).toHaveLength(0);
  });
  test('should render meta tags properly', () => {
    const metaTags = shallow(<MetaTag {...metaData} />);
    expect(metaTags).toMatchSnapshot();
  });
});
