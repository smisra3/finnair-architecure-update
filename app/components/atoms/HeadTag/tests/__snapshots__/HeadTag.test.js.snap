// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeadTag Component should render correctly and match snapshot 1`] = `
<SideEffect(Head)>
  <title>
    title
  </title>
  <MetaTag
    content="author"
    key="meta-1"
    meta={
      Object {
        "id": 1,
        "key": "author",
      }
    }
  />
  <MetaTag
    content="description"
    key="meta-2"
    meta={
      Object {
        "id": 2,
        "key": "description",
      }
    }
  />
  <MetaTag
    content="title"
    key="meta-twitter-3"
    meta={
      Object {
        "id": 3,
        "key": "title",
        "prefix": Array [
          "twitter",
          "og",
        ],
      }
    }
    prefix="twitter"
  />
  <MetaTag
    content="title"
    key="meta-og-3"
    meta={
      Object {
        "id": 3,
        "key": "title",
        "prefix": Array [
          "twitter",
          "og",
        ],
      }
    }
    prefix="og"
  />
  <MetaTag
    content="description"
    key="meta-twitter-4"
    meta={
      Object {
        "id": 4,
        "key": "description",
        "prefix": Array [
          "twitter",
          "og",
        ],
      }
    }
    prefix="twitter"
  />
  <MetaTag
    content="description"
    key="meta-og-4"
    meta={
      Object {
        "id": 4,
        "key": "description",
        "prefix": Array [
          "twitter",
          "og",
        ],
      }
    }
    prefix="og"
  />
  <MetaTag
    content="url"
    key="meta-twitter-5"
    meta={
      Object {
        "id": 5,
        "key": "url",
        "prefix": Array [
          "twitter",
          "og",
        ],
      }
    }
    prefix="twitter"
  />
  <MetaTag
    content="url"
    key="meta-og-5"
    meta={
      Object {
        "id": 5,
        "key": "url",
        "prefix": Array [
          "twitter",
          "og",
        ],
      }
    }
    prefix="og"
  />
  <MetaTag
    content="image"
    key="meta-twitter-6"
    meta={
      Object {
        "id": 6,
        "key": "image",
        "prefix": Array [
          "twitter",
          "og",
        ],
      }
    }
    prefix="twitter"
  />
  <MetaTag
    content="image"
    key="meta-og-6"
    meta={
      Object {
        "id": 6,
        "key": "image",
        "prefix": Array [
          "twitter",
          "og",
        ],
      }
    }
    prefix="og"
  />
  <MetaTag
    content="sample_description"
    key="meta-twitter-7"
    meta={
      Object {
        "id": 7,
        "key": "card",
        "prefix": Array [
          "twitter",
        ],
      }
    }
    prefix="twitter"
  />
  <MetaTag
    content="type"
    key="meta-twitter-8"
    meta={
      Object {
        "id": 8,
        "key": "type",
        "prefix": Array [
          "twitter",
          "og",
        ],
      }
    }
    prefix="twitter"
  />
  <MetaTag
    content="type"
    key="meta-og-8"
    meta={
      Object {
        "id": 8,
        "key": "type",
        "prefix": Array [
          "twitter",
          "og",
        ],
      }
    }
    prefix="og"
  />
  <MetaTag
    content="siteName"
    key="meta-og-9"
    meta={
      Object {
        "id": 9,
        "key": "site_name",
        "prefix": Array [
          "og",
        ],
        "serverKey": "siteName",
      }
    }
    prefix="og"
  />
</SideEffect(Head)>
`;

exports[`HeadTag Component should render correctly when no props passed 1`] = `
<SideEffect(Head)>
  <title />
</SideEffect(Head)>
`;

exports[`HeadTag Component should render meta tags properly 1`] = `
<meta
  content="content"
  name="prefix:meta key"
/>
`;
